### DISCUSSION TITLE
504 Gateway Time-out for requests which takes more than a minute.

### DISCUSSION THREAD

**Discussion by iamakashk** [apiman/apiman#1518] [Help]
Hi 

I am using apiman v2.1.0 and vertx gateway. I have one service which takes more than 2 min to process data( I have few other services which takes more than 2 min ) . When I directly call that service using endpoint of the services it works fine. But similar thing when I try from vertx gateway it gives me **504 Gateway Time-out**. 

I read on the vertx documentation section that default timeout is 30 sec. ( https://vertx.io/docs/vertx-core/java/ ). Do we have same for apiman vertx gateway ? I tried to look for similar property to put into conf-es.json but could not find any relevant documentation for the same. 

How can we configure the value of this timeout ?
Can anyone help me out here on this? 

![image](https://user-images.githubusercontent.com/19202603/135458203-05dfe63b-1fd8-4aa2-88b0-3ff111c75e05.png)

**Comment by msavy** [comment]
I made a branch/patch recently that you participated in where you were able to alter connection options. I take it that didn't work for you?

**Reply by iamakashk** [reply]
Hi @msavy 
Sorry I dont remember which branch you are referring to. Could you please help me recall that?

**Reply by msavy** [reply]
https://github.com/apiman/apiman/discussions/1373#discussioncomment-980520

Sorry, I can't spare much time at the moment as I need to focus on paid client work.

**Reply by iamakashk** [reply]
Hey @msavy thank you for prompt reply. I totally understand your situation. But my issue is little bit different from the thread you mentioned. That thread was related to gateway blocking calls. Mine is different. For me if backend services takes more than a minute to respond then vertx gateway responding timeout. I want to add some configuration so that it will wait till the backend service responds. Even if backend service takes longer time to respond , vertx gateway should wait. Requesting you to spare some time for this . This would really help me.

**Reply by iamakashk** [reply]
@msavy 

As suggested [here ](https://github.com/apiman/apiman/discussions/1373#discussioncomment-980520)

I tried with following config file but still request is timing out after 60 sec ( default timeout ). This configuration is somehow not overriding default configuration of vertx.

```
  "connector-factory": {
    "class": "io.apiman.gateway.platforms.vertx3.connector.ConnectorFactory",
    "config": {
	   // Your usual config here
       "httpClient": {
         "maxPoolSize": 100,
	 "keepAlive":true,
	 "http2KeepAliveTimeout":300000,
	 "keepAliveTimeout":300000,
	 "connectTimeout":300000,
	 "idleTimeout":300000,
	 "webSocketClosingTimeout":300000
       },
       }
```

 You have mentioned here that it wont make it in 2.1 release. Just wondering if I am doing right steps or not as I am still on 2.1 release.  Could you please help me out in the right direction here ?

**Reply by iamakashk** [reply]
update : 
I tried creating a jar of https://github.com/msavy/apiman/tree/allow-custom-client-options . But still request times out in 1 min.

**Reply by msavy** [reply]
Thank you for your comments.

Paid client work is what keeps the Apiman project thriving. As with all software engineering professionals, I have to make money to survive. Hence, those supporting the project get priority. 

I will try to provide answers when I have a few spare moments. But at the moment I'm exceptionally busy, so you may have to wait a long time and I can not guarantee anything.

Have you tried the usual method of overriding the timeouts? Here is an example to try (use it along with the other options you wanted to set).

```json
  "connector-factory": {
    "class": "io.apiman.gateway.platforms.vertx3.connector.ConnectorFactory",
    "config": {
        "http": {
            "timeouts": {
                "read": 123456,
                "connect": 123456
            }
        }
    }
  }
```

And yes, you need to build the branch I linked to manually.

**Reply by iamakashk** [reply]
@msavy Thanks for your response. I actually hardcoded those values in the code itself to make this work. I used the branch you shared. Currently its working fine for me. Although its not the nicest approach but for now its working. I would definitely check out with the option you have suggested and update on the same. Thanks a lot for right direction. 🙌 and Keep up the good work!!
